### ISSUE TITLE
Running TimeEval in cluster mode 

### ISSUE THREAD

**Issue by B-Seif** [HPI-Information-Systems/TimeEval#6]
I have followed the different steps to use TimeEval in cluster mode and I am having some issues. I have two machines in my cluster. The driver (32 cores and 32 G M) and a worker (16 cores and 32 G M).
When I run the script `eval.py` : 

```python
#!/usr/bin/env python3
import timeeval
assert timeeval.__version__ == "1.2.4", "TimeEval version not supported. This script is for TimeEval version 1.2.4!"
from pathlib import Path
from timeeval import TimeEval, MultiDatasetManager, Metric, Algorithm, TrainingType, InputDimensionality, ResourceConstraints
from timeeval.adapters import DockerAdapter
from timeeval.params import FixedParameters
from timeeval.resource_constraints import GB
from timeeval.params import FullParameterGrid
from timeeval.utils.window import ReverseWindowing
from timeeval import TimeEval, RemoteConfiguration, ResourceConstraints
import numpy as np

# post-processing 
def post_algo_custom(scores: np.ndarray, args: dict) -> np.ndarray:
    window_size = args.get("hyper_params", {}).get("window_size", 20)
    return ReverseWindowing(window_size=window_size).fit_transform(scores)

def main():
    # cluster setup
    cluster_config = RemoteConfiguration(
    scheduler_host="192.168.100.57",
    worker_hosts=["192.168.100.57","192.168.100.115"]
     )
    dm = MultiDatasetManager([Path("/home/ubuntu/TimeEval-test/tests/example_data")])
    datasets = dm.select() 
    algorithms = [
        Algorithm(
            name= "algo_custom",
            main=DockerAdapter(image_name="registry.gitlab.hpi.de/akita/i/algo_custom",tag="latest", skip_pull=True),
            data_as_file=True,
            param_config=FullParameterGrid({
                "window_size" : [50],
                "lamda" : [0.1],
                "beta"  :  [0.1]
                }),
                postprocess=post_algo_custom,
            training_type=TrainingType.UNSUPERVISED,
            input_dimensionality=InputDimensionality.MULTIVARIATE
        )
        ]
    repetitions = 5
    #rcs = ResourceConstraints.default_constraints()
    limits = ResourceConstraints(tasks_per_host=16)

    timeeval = TimeEval(
        dm, 
        datasets, 
        algorithms,
        repetitions=repetitions, 
        metrics=[Metric.ROC_AUC,Metric.RANGE_PR_AUC],
        remote_config=cluster_config,
        resource_constraints=limits,
        distributed=True
    )
    timeeval.run()
    results = timeeval.get_results(aggregated=True)
    print(results)

if __name__ == "__main__":
    main()
```

I get this worker error and I really don't know how to debug this error, yet both machines have exactly the same setup (docker, python 3.8.8, dask,..)
`Exception TypeError('Metric() takes no arguments') occurred remotely.` 

In the worker log file I have this:
```python
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/worker.py", line 2692, in _maybe_deserialize_task
    function, args, kwargs = _deserialize(*ts.runspec)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/worker.py", line 3545, in _deserialize
    function = loads_function(function)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/worker.py", line 3536, in loads_function
    result = pickle.loads(bytes_object)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
```

```
Exception TypeError('Metric() takes no arguments') occurred remotely.
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/timeeval/timeeval.py", line 223, in get_future_result
    r = f.result()
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
Exception TypeError('Metric() takes no arguments') occurred remotely.
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/timeeval/timeeval.py", line 223, in get_future_result
    r = f.result()
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
Exception TypeError('Metric() takes no arguments') occurred remotely.
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/timeeval/timeeval.py", line 223, in get_future_result
    r = f.result()
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
Exception TypeError('Metric() takes no arguments') occurred remotely.
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/timeeval/timeeval.py", line 223, in get_future_result
    r = f.result()
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
Exception TypeError('Metric() takes no arguments') occurred remotely.
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/timeeval/timeeval.py", line 223, in get_future_result
    r = f.result()
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
TypeError: Metric() takes no arguments
Running FINALIZE phase
Finalizing: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.01s/it]
Collecting results: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.27s/it]
FINALIZE phase done.
          Stored results at /home/ubuntu/TimeEval-test/results/2022_07_11_14_08_17/results.csv.
          Overall runtime of this TimeEval run: 6.330169200897217 seconds
        
                                     ROC_AUC_mean  RANGE_PR_AUC_mean  train_main_time_mean  execute_main_time_mean  repetitions
algorithm collection dataset                                                                                                   
algo_custom  CalIt2     CalIt2-traffic           NaN                NaN                   NaN                     NaN            5
```

**Comment by CodeLionX** [comment]
Please always use the same TimeEval-version on all cluster nodes.

Regarding your second issue, I can't see any issue.
The Dask-logs indicate that the Dask cluster successfully starts up and is now waiting for a client connection / jobs. Are you confident that all nodes can reach each other, your driver process can reach the scheduler, and no firewall is blocking the requests? Have you seen any error message in the driver process (e.g., `timeeval.log`)?

**Comment by B-Seif** [comment]
Even using only the localhost (driver+worker), the cluster remains blocked, so I don't think it would be a network problem? what do I need to look at?
```python
# cluster setup
    cluster_config = RemoteConfiguration(
    scheduler_host="localhost", 
    #worker_hosts=["192.168.100.57"] #"192.168.100.115"
     )

```

```
distributed.deploy.ssh - INFO - distributed.scheduler - INFO - -----------------------------------------------
distributed.deploy.ssh - INFO - numexpr.utils - INFO - Note: NumExpr detected 32 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
distributed.deploy.ssh - INFO - numexpr.utils - INFO - NumExpr defaulting to 8 threads.
distributed.deploy.ssh - INFO - distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
distributed.deploy.ssh - INFO - bokeh.server.util - WARNING - Host wildcard '*' will allow connections originating from multiple (or possibly all) hostnames or IPs. Use non-wildcard values to restrict access explicitly
distributed.deploy.ssh - INFO - bokeh.server.tornado - INFO - Keep-alive ping configured every 500 milliseconds
distributed.deploy.ssh - INFO - bokeh.server.tornado - INFO - Check for unused sessions every 500 milliseconds
distributed.deploy.ssh - INFO - bokeh.server.tornado - INFO - User authentication hooks NOT provided (default user enabled)
distributed.deploy.ssh - INFO - distributed.scheduler - INFO - -----------------------------------------------
distributed.deploy.ssh - INFO - distributed.scheduler - INFO - Clear task state
distributed.deploy.ssh - INFO - distributed.scheduler - INFO -   Scheduler at: tcp://192.168.100.57:8786
Running PREPARE phase
Preparing: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 239.73it/s]
Running EVALUATION phase
Submitting evaluation tasks: 100%|███████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 126.77it/s]
Evaluating distributedly:   0%|                                                                                               | 0/5 [00:34<?, ?it/s]

```

**Comment by CodeLionX** [comment]
> ```python
> # cluster setup
> cluster_config = RemoteConfiguration(
>     scheduler_host="localhost", 
>     # worker_hosts=["192.168.100.57"] #"192.168.100.115"
> )
> ```

This can't work. You need to at least have one worker; otherwise there is no entity to process any tasks/jobs.

Please familiarize yourself with Dask and [how to deploy a Dask cluster](https://docs.dask.org/en/stable/deploying.html). We use the term _driver_ for the Python process that runs the Dask `Client`.

**Comment by B-Seif** [comment]
By rebooting the machines; it ended up working! but it's a bit slow, maybe because I didn't configure the cluster well.

So I have a few additional questions:

- Here is the final composition of my cluster (hardware) and  I would like to know if you could recommend me an optimal configuration (software) : 
 4 machines (32cores/32 Giga, 16 Cores/32Giga, 8/Cores/16Giga, 8Cores/16 Giga). 
Which to take as a driver? how many tasks per host?

- Another problem concerning the docker images: I create the images by hand in each worker ! and I would like to make sure that workers can download images from the registry (by using the flag skip_pull= False). However, by doing this, I have an error accessing your registry.

 For example instead of creating the image of the` lof `in each worker, I would like to create it only once, put it in a registry and then set the flag `skip_pull=False` in the driver, and like that the workers can download the `lof` image  from the registry `registry.gitlab.hpi.de/akita/i/lof`.

**Comment by CodeLionX** [comment]
I'm glad you got it working. 👍🏼

Yes, the Docker registry `registry.gitlab.hpi.de` is just available for HPI students and staff. Hosting the algorithm Docker images publicly is – for now – out of scope of this project. If there will be more demand for them, we as a community can think about a solution for that.

Your additional questions are out of scope of TimeEval and this issue. There might be multiple answers to them. Please open new discussions for them [here](https://github.com/HPI-Information-Systems/TimeEval/discussions).
